src/core/lwt_result.ml

```diff
@@ -104,6 +104,29 @@ let iter_error f r =
       | Error e -> f e
       | Ok _ -> Lwt.return_unit)
 
+let partition_filter_map f l =
+  let rec iter f l1 l2 l =
+    match l with
+    | [] -> (List.rev l1, List.rev l2)
+    | x :: tl -> (
+        match f x with
+        | `Left y -> iter f (y :: l1) l2 tl
+        | `Right y -> iter f l1 (y :: l2) tl
+        | `Drop -> iter f l1 l2 tl)
+  in
+  iter f [] [] l
+
+let split_result results =
+  partition_filter_map
+    (fun x -> match x with Ok o -> `Left o | Error e -> `Right e)
+    results
+
+let collect x =
+  x |> Lwt.all
+  |> Lwt.map (fun r ->
+          let oks, errors = split_result r in
+          match errors with [] -> Ok oks | _ -> Error errors)
```

Comment:
On a quite specific note: This seems like a simple yet costly way to implement collect. E.g., the `oks` list is reversed (line 110) whether there are errors or not.

On a more general note: the fact that the Lwt part and the result part of this function are entirely independent makes me wonder whether it actually belongs in Lwt.

Reply:
Thank you for the feedback.

> This seems like a simple yet costly way to implement collect

You're right. If we wanted to add `collect` to `Lwt_result`, then I'd be happy to optimize this.

> the fact that the Lwt part and the result part of this function are entirely independent makes me wonder whether it actually belongs in Lwt

You're also right that the Lwt and result parts are distinct. If, for example, `collect` were in `CCResult` as 

```ocaml
let collect x =
  let oks, errors = split_result x in
  match errors with [] -> Ok oks | _ -> Error errors
```

then usage might look like

```ocaml
my_data_list
|> request_for_each
|> Lwt.all |> Lwt.map CCResult.collect
```

which isn't much different from or less convenient than

```ocaml
my_data_list
|> request_for_each
|> Lwt_result.collect
```

If you think that, even with a more efficient implementation of `result list` filtering and partitioning, this code does not belong in Lwt, then I'd be happy to close this PR.